File: scripts/test_aarch64_block_owner.py
#!/usr/bin/env python3
"""Run production block queue/lock code with host counter/event/MMIO adapters."""

from __future__ import annotations

import pathlib
import re
import subprocess
import tempfile

ROOT = pathlib.Path(__file__).resolve().parents[1]
BLOCK = (ROOT / "kernel/src/aarch64_virtio_blk.rs").read_text()
VFS = (ROOT / "kernel/src/vfs.rs").read_text()
VOLUME = (ROOT / "kernel/src/makfs4_volume.rs").read_text()


def item(source: str, start: str) -> str:
    offset = source.index(start)
    opening = source.index("{", offset)
    depth = 1
    end = opening + 1
    while depth:
        depth += (source[end] == "{") - (source[end] == "}")
        end += 1
    return source[offset:end]


def lock_item(source: str, start: str) -> str:
    result = item(source, start)
    # Select the actual AArch64 branch for this host-side execution, including
    # on Darwin/x86. No lock logic or request-service call is replaced.
    return result.replace('#[cfg(target_arch = "aarch64")]\n', "")


constants = "\n".join(
    re.findall(r"^const (?:MAX_DEVICES|REQUEST_(?:READ|WRITE|FLUSH)|SERVICE_\w+|SLOT_\w+):.*;$", BLOCK, re.M)
)
counts = "\n".join(
    re.findall(r"^static (?:NONOWNER_REQUESTS|OWNER_\w+|TIMER_SERVICE_COMPLETIONS):.*;$", BLOCK, re.M)
)
queue_types = BLOCK[
    BLOCK.index("#[derive(Clone, Copy)]\nstruct ServiceRequest") :
    BLOCK.index("pub struct SourceWriteFreeze")
]
functions = "\n".join(
    item(BLOCK, declaration)
    for declaration in (
        "fn queue_request(",
        "pub fn service_requests_from_timer()",
        "pub fn service_requests_while_waiting()",
        "fn service_requests(timer_service:",
        "pub fn reset_service_affinity_evidence()",
        "pub fn service_affinity_evidence()",
    )
)
assert "counter_deadline_millis(5_000)" in functions
assert "wait_for_service_event();" in functions
assert "notify_service_waiters();" in functions
assert '"dsb ish", "sev"' in item(BLOCK, "fn notify_service_waiters()")
assert '"wfe"' in item(BLOCK, "fn wait_for_service_event()")

lock_functions = {
    "vfs_lock.rs": lock_item(VFS, "fn with_state<R>"),
    "inode_lock.rs": lock_item(VOLUME, "fn with_inode_cache<R>"),
    "mutation_lock.rs": "struct MutationGuard;\n"
    + lock_item(VOLUME, "impl MutationGuard")
    + "\n"
    + item(VOLUME, "impl Drop for MutationGuard"),
}
for name, body in lock_functions.items():
    assert "crate::aarch64_virtio_blk::service_requests_while_waiting();" in body, name
    assert "compare_exchange_weak" in body and "Ordering::Acquire" in body, name

with tempfile.TemporaryDirectory(prefix="makos-block-owner-") as directory:
    output = pathlib.Path(directory)
    (output / "block_queue.rs").write_text(
        "use core::cell::UnsafeCell;\n"
        "use core::sync::atomic::{AtomicBool, AtomicU8, AtomicU64, Ordering};\n"
        + constants + "\n" + counts + "\n" + queue_types + "\n" + functions
    )
    for name, body in lock_functions.items():
        (output / name).write_text(body)
    # include! resolves the unmodified behavioral fixture beside these exact
    # source excerpts. Hardware access alone is supplied by the host fixture.
    (output / "test.rs").write_text((ROOT / "scripts/test_aarch64_block_owner.rs").read_text())
    binary = output / "block-owner-test"
    subprocess.run(
        ["rustc", "--edition=2024", "--test", str(output / "test.rs"), "-o", str(binary)],
        check=True,
    )
    subprocess.run([str(binary), "--test-threads=1"], check=True, timeout=30)

    # Remove only the new VFS progress call in a generated negative control.
    # The exact same behavioral regression must then observe the circular
    # wait, demonstrating that ordinary host scheduling cannot hide the bug.
    (output / "vfs_lock.rs").write_text(
        lock_functions["vfs_lock.rs"].replace(
            "crate::aarch64_virtio_blk::service_requests_while_waiting();", ""
        )
    )
    negative = output / "block-owner-without-vfs-progress"
    subprocess.run(
        ["rustc", "--edition=2024", "--test", str(output / "test.rs"), "-o", str(negative)],
        check=True,
    )
    failed = subprocess.run(
        [str(negative), "--exact", "vfs_contention_completes_ap_io_without_timer"],
        capture_output=True, text=True, timeout=15,
    )
    if failed.returncode == 0 or "CPU0/AP filesystem-block circular wait" not in failed.stdout + failed.stderr:
        raise SystemExit(f"block owner negative control did not reproduce circular wait:\n{failed.stdout}{failed.stderr}")

print("MAKOS_AARCH64_BLOCK_OWNER_HOST_OK locks=vfs,inode-cache,mutation "
      "service=production-queue timer=absent-in-contention requests=read,write,flush "
      "owner=cpu0 timeout_ms=5000 negative_control=deadlock-reproduced")

File: scripts/test_aarch64_block_owner.rs
//! Host hardware adapters for exact production queue and filesystem lock code.

use std::cell::Cell;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Arc, Barrier, mpsc};
use std::thread;
use std::time::{Duration, Instant};

mod arch {
    use super::*;
    thread_local! { static CPU: Cell<usize> = const { Cell::new(0) }; }
    static CLOCK: std::sync::OnceLock<Instant> = std::sync::OnceLock::new();
    pub static DEADLINES: AtomicU64 = AtomicU64::new(0);
    pub fn cpu_index() -> usize {
        CPU.get()
    }
    pub fn set_cpu(cpu: usize) {
        CPU.set(cpu);
    }
    fn now() -> u64 {
        CLOCK.get_or_init(Instant::now).elapsed().as_millis() as u64
    }
    pub fn counter_deadline_millis(millis: u64) -> u64 {
        assert_eq!(
            millis, 5_000,
            "production request deadline must remain unchanged"
        );
        DEADLINES.fetch_add(1, Ordering::AcqRel);
        now() + millis
    }
    pub fn counter_deadline_expired(deadline: u64) -> bool {
        now() >= deadline
    }
}

fn fatal(message: &str) -> ! {
    panic!("{message}")
}

mod aarch64_virtio_blk {
    include!("block_queue.rs");

    // No timer thread runs in this test: progress must come from the CPU0
    // contender executing the production filesystem lock path.
    struct LockedState {
        lock: AtomicBool,
    }
    static STATE: LockedState = LockedState {
        lock: AtomicBool::new(false),
    };
    static DEVICE_OPERATIONS: AtomicU64 = AtomicU64::new(0);
    fn notify_service_waiters() {
        std::thread::yield_now();
    }
    fn wait_for_service_event() {
        std::thread::yield_now();
    }
    fn device_operation(action: impl FnOnce() -> bool) -> bool {
        assert_eq!(crate::arch::cpu_index(), 0, "non-owner touched device");
        assert!(
            STATE
                .lock
                .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
                .is_ok()
        );
        DEVICE_OPERATIONS.fetch_add(1, Ordering::AcqRel);
        let result = action();
        STATE.lock.store(false, Ordering::Release);
        result
    }
    fn byte(lba: u32, index: usize) -> u8 {
        (lba as usize + index) as u8
    }
    fn read_on_owner(device: usize, lba: u32, output: &mut [u8]) -> bool {
        device_operation(|| {
            assert_eq!(device, 1);
            for (index, value) in output.iter_mut().enumerate() {
                *value = byte(lba, index);
            }
            true
        })
    }
    fn write_on_owner(device: usize, lba: u32, input: &[u8]) -> bool {
        device_operation(|| {
            assert_eq!(device, 1);
            assert!(
                input
                    .iter()
                    .enumerate()
                    .all(|(index, value)| *value == byte(lba, index))
            );
            true
        })
    }
    fn flush_on_owner(device: usize) -> bool {
        device_operation(|| {
            assert_eq!(device, 1);
            true
        })
    }
    pub fn read(lba: u32, length: usize) {
        let mut bytes = vec![0u8; length];
        assert!(queue_request(REQUEST_READ, 1, lba, None, Some(&mut bytes)));
        assert!(
            bytes
                .iter()
                .enumerate()
                .all(|(index, value)| *value == byte(lba, index))
        );
    }
    pub fn write(lba: u32, length: usize) {
        let bytes: Vec<_> = (0..length).map(|index| byte(lba, index)).collect();
        assert!(queue_request(REQUEST_WRITE, 1, lba, Some(&bytes), None));
    }
    pub fn flush() {
        assert!(queue_request(REQUEST_FLUSH, 1, 0, None, None));
    }
    pub fn pending() -> usize {
        SERVICE
            .iter()
            .filter(|slot| slot.state.load(Ordering::Acquire) == SLOT_READY)
            .count()
    }
    pub fn operations() -> u64 {
        DEVICE_OPERATIONS.load(Ordering::Acquire)
    }
    pub fn lock_device(locked: bool) {
        STATE.lock.store(locked, Ordering::Release);
    }
    pub fn reset() {
        assert!(
            SERVICE
                .iter()
                .all(|slot| slot.state.load(Ordering::Acquire) == SLOT_FREE)
        );
        assert!(!STATE.lock.load(Ordering::Acquire));
        reset_service_affinity_evidence();
        DEVICE_OPERATIONS.store(0, Ordering::Release);
        crate::arch::DEADLINES.store(0, Ordering::Release);
    }
}

mod vfs {
    use core::cell::UnsafeCell;
    use core::sync::atomic::{AtomicBool, Ordering};
    type State = u64;
    struct LockedState {
        lock: AtomicBool,
        state: UnsafeCell<State>,
    }
    unsafe impl Sync for LockedState {}
    static STATE: LockedState = LockedState {
        lock: AtomicBool::new(false),
        state: UnsafeCell::new(0),
    };
    include!("vfs_lock.rs");
    pub fn run(action: impl FnOnce()) {
        with_state(|_| action());
    }
}

mod volume {
    use core::cell::UnsafeCell;
    use core::sync::atomic::{AtomicBool, Ordering};
    type InodeCache = u64;
    struct LockedCache {
        lock: AtomicBool,
        value: UnsafeCell<InodeCache>,
    }
    unsafe impl Sync for LockedCache {}
    static INODE_CACHE: LockedCache = LockedCache {
        lock: AtomicBool::new(false),
        value: UnsafeCell::new(0),
    };
    static MUTATION_LOCK: AtomicBool = AtomicBool::new(false);
    include!("inode_lock.rs");
    include!("mutation_lock.rs");
    pub fn cache(action: impl FnOnce()) {
        with_inode_cache(|_| action());
    }
    pub fn mutate(action: impl FnOnce()) {
        let _guard = MutationGuard::acquire();
        action();
    }
}

fn await_pending(count: usize) {
    let start = Instant::now();
    while aarch64_virtio_blk::pending() != count {
        assert!(
            start.elapsed() < Duration::from_secs(2),
            "AP did not publish request"
        );
        thread::yield_now();
    }
}

#[derive(Clone, Copy)]
enum Lock {
    Vfs,
    Cache,
    Mutation,
}
fn locked(lock: Lock, action: impl FnOnce()) {
    match lock {
        Lock::Vfs => vfs::run(action),
        Lock::Cache => volume::cache(action),
        Lock::Mutation => volume::mutate(action),
    }
}

fn contended_lock_progress(lock: Lock) {
    aarch64_virtio_blk::reset();
    let (acquired, acquisition) = mpsc::channel();
    let ap = thread::spawn(move || {
        arch::set_cpu(1);
        locked(lock, || {
            acquired.send(()).unwrap();
            aarch64_virtio_blk::read(51, 4096);
            aarch64_virtio_blk::write(72, 512);
            aarch64_virtio_blk::flush();
        });
    });
    acquisition.recv_timeout(Duration::from_secs(2)).unwrap();
    await_pending(1);
    assert_eq!(aarch64_virtio_blk::operations(), 0);
    let (finished, completion) = mpsc::channel();
    let owner = thread::spawn(move || {
        arch::set_cpu(0);
        locked(lock, || finished.send(()).unwrap());
    });
    // This is a host regression bound, not a modified guest timeout. Without
    // the contention hook CPU0 cannot acquire the lock and the AP stays READY.
    completion
        .recv_timeout(Duration::from_secs(2))
        .expect("CPU0/AP filesystem-block circular wait");
    ap.join().unwrap();
    owner.join().unwrap();
    assert_eq!(
        aarch64_virtio_blk::service_affinity_evidence(),
        (3, 3, 1, 1, 1, 0)
    );
    assert_eq!(arch::DEADLINES.load(Ordering::Acquire), 3);
}

#[test]
fn vfs_contention_completes_ap_io_without_timer() {
    contended_lock_progress(Lock::Vfs);
}
#[test]
fn inode_cache_contention_completes_ap_io_without_timer() {
    contended_lock_progress(Lock::Cache);
}
#[test]
fn mutation_contention_completes_ap_io_without_timer() {
    contended_lock_progress(Lock::Mutation);
}

#[test]
fn three_concurrent_requests_preserve_identity_and_owner_affinity() {
    aarch64_virtio_blk::reset();
    let barrier = Arc::new(Barrier::new(4));
    let mut aps = Vec::new();
    for cpu in 1..=3 {
        let barrier = barrier.clone();
        aps.push(thread::spawn(move || {
            arch::set_cpu(cpu);
            barrier.wait();
            match cpu {
                1 => aarch64_virtio_blk::read(9, 512),
                2 => aarch64_virtio_blk::write(17, 4096),
                _ => aarch64_virtio_blk::flush(),
            }
        }));
    }
    barrier.wait();
    await_pending(3);
    arch::set_cpu(1);
    assert_eq!(aarch64_virtio_blk::service_requests_while_waiting(), 0);
    assert_eq!(aarch64_virtio_blk::operations(), 0);
    arch::set_cpu(0);
    assert_eq!(aarch64_virtio_blk::service_requests_while_waiting(), 3);
    for ap in aps {
        ap.join().unwrap();
    }
    assert_eq!(
        aarch64_virtio_blk::service_affinity_evidence(),
        (3, 3, 1, 1, 1, 0)
    );
}

#[test]
fn recursive_device_entry_defers_and_timer_service_remains_available() {
    aarch64_virtio_blk::reset();
    let ap = thread::spawn(|| {
        arch::set_cpu(1);
        aarch64_virtio_blk::read(97, 512);
    });
    await_pending(1);
    arch::set_cpu(0);
    aarch64_virtio_blk::lock_device(true);
    assert_eq!(aarch64_virtio_blk::service_requests_while_waiting(), 0);
    assert_eq!(aarch64_virtio_blk::service_requests_from_timer(), 0);
    assert_eq!(aarch64_virtio_blk::pending(), 1);
    assert_eq!(aarch64_virtio_blk::operations(), 0);
    aarch64_virtio_blk::lock_device(false);
    assert_eq!(aarch64_virtio_blk::service_requests_from_timer(), 1);
    ap.join().unwrap();
    assert_eq!(
        aarch64_virtio_blk::service_affinity_evidence(),
        (1, 1, 1, 0, 0, 1)
    );
}
